nym-vpn-app/src-tauri/src/grpc/tunnel.rs

```diff
@@ -0,0 +1,264 @@
+use serde::Serialize;
+use time::OffsetDateTime;
+use ts_rs::TS;
+
+use nym_vpn_proto as p;
+use p::tunnel_state::{ActionAfterDisconnect, ErrorStateReason, State};
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct WgNode {
+    pub endpoint: String,
+    pub public_key: String,
+    pub private_ipv4: String,
+    pub private_ipv6: String,
+}
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct MixnetData {
+    pub nym_address: Option<String>,
+    pub exit_ipr: Option<String>,
+    pub ipv4: String,
+    pub ipv6: String,
+}
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct WireguardData {
+    pub entry: WgNode,
+    pub exit: WgNode,
+}
+
+#[derive(Serialize, Clone, Debug, PartialEq, TS)]
+#[ts(export)]
+#[serde(untagged)]
+pub enum TunnelData {
+    Mixnet(MixnetData),
+    Wireguard(WireguardData),
+}
+
+#[derive(Serialize, Clone, PartialEq, Debug, TS)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub struct Tunnel {
+    pub entry_gw_id: String,
+    pub exit_gw_id: String,
+    pub connected_at: Option<i64>,
+    pub data: TunnelData,
+}
+
+#[derive(Default, Debug, Clone, Serialize, PartialEq, TS, strum::Display)]
+#[ts(export)]
+#[serde(rename_all = "camelCase")]
+pub enum TunnelState {
+    #[default]
+    Disconnected,
+    Connected(Tunnel),
+    Connecting(Option<Tunnel>),
+    Disconnecting(Option<TunnelAction>),
+    Error(TunnelError),
+    Offline {
+        reconnect: bool,
+    },
+}
+
+impl TunnelState {
+    pub fn from_proto(tunnel: &State) -> Result<TunnelState, &str> {
```

Comment:
Why not implement `From` trait instead?

Reply:
Why not using a simple method instead of trait implem?

Reply:
Well because `From` and `TryFrom` traits are somewhat standardised way to perform conversions. Also consistency, it's hard for me to say why sometimes you prefer From/TryFrom vs `from_proto()`